File: Windows/PasswallReceiver/Core/FileTransferBatch.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace PasswallReceiver.Core;

internal sealed record FileTransferBatch(FileTransferManifest Manifest, IReadOnlyList<string> Files)
{
    public static FileTransferBatch Build(
        IEnumerable<string> selectedPaths,
        ulong maximumBatchBytes = ProtocolContract.MaximumBatchBytes)
    {
        maximumBatchBytes = Math.Min(maximumBatchBytes, ProtocolContract.MaximumBatchBytes);
        var wireEntries = new List<Dictionary<string, object?>>();
        var files = new List<string>();
        ulong totalBytes = 0;
        var discoveredEntries = 0;
        foreach (var path in selectedPaths)
        {
            Append(
                Path.GetFullPath(path),
                Path.GetFileName(path),
                wireEntries,
                files,
                ref totalBytes,
                ref discoveredEntries,
                maximumBatchBytes);
        }
        var document = JsonSerializer.SerializeToElement(new { entries = wireEntries });
        return new FileTransferBatch(FileTransferManifest.Parse(document), files);
    }

    public Stream OpenRead() => new FileTransferSourceStream(this);

    private static void Append(
        string path,
        string relativePath,
        List<Dictionary<string, object?>> entries,
        List<string> files,
        ref ulong totalBytes,
        ref int discoveredEntries,
        ulong maximumBatchBytes,
        bool alreadyDiscovered = false)
    {
        if (!alreadyDiscovered)
        {
            if (discoveredEntries == ProtocolContract.MaximumBatchEntries)
            {
                throw new InvalidDataException("File batch exceeds the entry limit");
            }
            discoveredEntries++;
        }
        var attributes = File.GetAttributes(path);
        if ((attributes & FileAttributes.ReparsePoint) != 0)
        {
            throw new InvalidDataException($"Symbolic links are unsupported: {relativePath}");
        }
        if ((attributes & FileAttributes.Directory) != 0)
        {
            entries.Add(new() { ["path"] = relativePath, ["kind"] = "directory", ["byteCount"] = 0UL });
            var remainingEntries = ProtocolContract.MaximumBatchEntries - discoveredEntries;
            var children = Directory.EnumerateFileSystemEntries(path)
                .Take(remainingEntries + 1)
                .ToArray();
            if (children.Length > remainingEntries)
            {
                throw new InvalidDataException("File batch exceeds the entry limit");
            }
            discoveredEntries += children.Length;
            foreach (var child in children.OrderBy(
                child => Path.GetFileName(child),
                StringComparer.Ordinal))
            {
                Append(
                    child,
                    $"{relativePath}/{Path.GetFileName(child)}",
                    entries,
                    files,
                    ref totalBytes,
                    ref discoveredEntries,
                    maximumBatchBytes,
                    alreadyDiscovered: true);
            }
            return;
        }
        if ((attributes & (FileAttributes.Device | FileAttributes.Offline)) != 0)
        {
            throw new InvalidDataException($"Unsupported file: {relativePath}");
        }
        var size = checked((ulong)new FileInfo(path).Length);
        try { totalBytes = checked(totalBytes + size); }
        catch (OverflowException error)
        {
            throw new InvalidDataException("File batch size overflowed", error);
        }
        if (totalBytes > maximumBatchBytes)
        {
            throw new InvalidDataException("File batch exceeds the byte limit");
        }
        using var source = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read);
        var digest = Convert.ToHexString(SHA256.HashData(source)).ToLowerInvariant();
        entries.Add(new()
        {
            ["path"] = relativePath,
            ["kind"] = "file",
            ["byteCount"] = size,
            ["sha256"] = digest
        });
        files.Add(path);
    }
}

internal sealed class FileTransferSourceStream(FileTransferBatch batch) : Stream
{
    private readonly List<FileManifestEntry> files = batch.Manifest.Entries
        .Where(entry => entry.Kind == "file").ToList();
    private int index;
    private FileStream? source;
    private ulong byteCount;
    private IncrementalHash hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);

    public override bool CanRead => true;
    public override bool CanSeek => false;
    public override bool CanWrite => false;
    public override long Length => checked((long)batch.Manifest.TotalBytes);
    public override long Position { get; set; }

    public override int Read(byte[] buffer, int offset, int count)
    {
        while (index < files.Count)
        {
            Prepare();
            var read = source!.Read(buffer, offset, count);
            if (read > 0)
            {
                Accept(buffer.AsSpan(offset, read));
                Position += read;
                return read;
            }
            FinishFile();
        }
        return 0;
    }

    public override async ValueTask<int> ReadAsync(
        Memory<byte> buffer,
        CancellationToken cancellationToken = default)
    {
        while (index < files.Count)
        {
            Prepare();
            var read = await source!.ReadAsync(buffer, cancellationToken);
            if (read > 0)
            {
                Accept(buffer.Span[..read]);
                Position += read;
                return read;
            }
            FinishFile();
        }
        return 0;
    }

    public override void Flush() { }
    public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
    public override void SetLength(long value) => throw new NotSupportedException();
    public override void Write(byte[] buffer, int offset, int count) => throw new NotSupportedException();

    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            source?.Dispose();
            hash.Dispose();
        }
        base.Dispose(disposing);
    }

    private void Prepare()
    {
        if (source is not null) return;
        source = new FileStream(batch.Files[index], FileMode.Open, FileAccess.Read, FileShare.Read);
        byteCount = 0;
        hash.Dispose();
        hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
    }

    private void Accept(ReadOnlySpan<byte> data)
    {
        byteCount += checked((ulong)data.Length);
        if (byteCount > files[index].ByteCount)
        {
            throw new InvalidDataException($"File changed after offer: {files[index].Path}");
        }
        hash.AppendData(data);
    }

    private void FinishFile()
    {
        source!.Dispose();
        source = null;
        var digest = Convert.ToHexString(hash.GetHashAndReset()).ToLowerInvariant();
        if (byteCount != files[index].ByteCount || digest != files[index].Sha256)
        {
            throw new InvalidDataException($"File changed after offer: {files[index].Path}");
        }
        index++;
    }
}

internal enum FileTransferStatus
{
    Queued,
    AwaitingApproval,
    Transferring,
    Verifying,
    Completed,
    Rejected,
    Canceled,
    Failed
}

internal sealed record FileTransferNameMapping(string Original, string Local);

internal sealed record FileTransferHistoryEntry(
    Guid TransferID,
    string Name,
    TransferDirection Direction,
    ulong TotalBytes,
    string DeviceName,
    DateTimeOffset StartedAt,
    FileTransferStatus Status,
    IReadOnlyList<FileTransferNameMapping>? NameMappings = null);

internal sealed class FileTransferHistoryStore
{
    public const int MaximumCount = 100;
    private readonly List<FileTransferHistoryEntry> entries = [];
    public IReadOnlyList<FileTransferHistoryEntry> Entries => entries;

    public void Record(FileTransferHistoryEntry entry)
    {
        entries.RemoveAll(existing => existing.TransferID == entry.TransferID);
        entries.Insert(0, entry);
        if (entries.Count > MaximumCount) entries.RemoveAt(entries.Count - 1);
    }

    public void Update(Guid transferID, FileTransferStatus status)
    {
        var index = entries.FindIndex(entry => entry.TransferID == transferID);
        if (index >= 0) entries[index] = entries[index] with { Status = status };
    }

    public void Clear() => entries.Clear();
}

File: Sources/PasswallCore/FileTransferBatch.swift
import CryptoKit
import Foundation

public struct FileTransferBatch: Sendable {
    public let manifest: FileTransferManifest
    public let fileURLs: [URL]

    public static func build(from selectedURLs: [URL]) throws -> Self {
        var entries: [FileManifestEntry] = []
        var files: [URL] = []
        var totalBytes: UInt64 = 0
        for url in selectedURLs {
            try Task.checkCancellation()
            try append(
                url,
                path: url.lastPathComponent,
                entries: &entries,
                files: &files,
                totalBytes: &totalBytes
            )
        }
        return try Self(manifest: FileTransferManifest(entries: entries), fileURLs: files)
    }

    private static func append(
        _ url: URL,
        path: String,
        entries: inout [FileManifestEntry],
        files: inout [URL],
        totalBytes: inout UInt64,
        enumerateDescendants: Bool = true
    ) throws {
        try Task.checkCancellation()
        guard entries.count < TransferLimits.maximumBatchEntries else {
            throw FileTransferManifestError.invalidEntryCount(entries.count + 1)
        }
        let values = try url.resourceValues(forKeys: [
            .isSymbolicLinkKey, .isDirectoryKey, .isRegularFileKey, .fileSizeKey
        ])
        guard values.isSymbolicLink != true else {
            throw FileTransferBatchError.unsupportedItem(path)
        }
        if values.isDirectory == true {
            entries.append(.directory(path: path))
            guard enumerateDescendants else { return }
            var enumerationError: Error?
            guard let enumerator = FileManager.default.enumerator(
                at: url,
                includingPropertiesForKeys: [.isSymbolicLinkKey, .isDirectoryKey, .isRegularFileKey],
                options: [],
                errorHandler: { _, error in
                    enumerationError = error
                    return false
                }
            ) else {
                throw FileTransferBatchError.unsupportedItem(path)
            }
            var children: [(url: URL, path: String)] = []
            while let child = enumerator.nextObject() as? URL {
                guard entries.count + children.count < TransferLimits.maximumBatchEntries else {
                    throw FileTransferManifestError.invalidEntryCount(
                        entries.count + children.count + 1
                    )
                }
                let relativePath = child.pathComponents
                    .suffix(enumerator.level)
                    .joined(separator: "/")
                children.append((child, "\(path)/\(relativePath)"))
            }
            if let enumerationError { throw enumerationError }
            for child in children.sorted(by: { $0.path < $1.path }) {
                try append(
                    child.url,
                    path: child.path,
                    entries: &entries,
                    files: &files,
                    totalBytes: &totalBytes,
                    enumerateDescendants: false
                )
            }
            return
        }
        guard values.isRegularFile == true, let size = values.fileSize, size >= 0 else {
            throw FileTransferBatchError.unsupportedItem(path)
        }
        let sum = totalBytes.addingReportingOverflow(UInt64(size))
        guard !sum.overflow, sum.partialValue <= TransferLimits.maximumBatchBytes else {
            throw FileTransferManifestError.payloadTooLarge
        }
        totalBytes = sum.partialValue
        entries.append(.file(
            path: path,
            byteCount: UInt64(size),
            sha256: try digest(url)
        ))
        files.append(url)
    }

    private static func digest(_ url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var hasher = SHA256()
        while try autoreleasepool(invoking: {
            guard
                let data = try handle.read(upToCount: BulkFrame.maximumChunkSize),
                !data.isEmpty
            else { return false }
            try Task.checkCancellation()
            hasher.update(data: data)
            return true
        }) {}
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }
}

public final class FileTransferBatchReader {
    private let batch: FileTransferBatch
    private let files: [FileManifestEntry]
    private let chunkSize: Int
    private var index = 0
    private var handle: FileHandle?
    private var bytesRead: UInt64 = 0
    private var hasher = SHA256()

    public init(batch: FileTransferBatch, chunkSize: Int = BulkFrame.maximumChunkSize) {
        self.batch = batch
        files = batch.manifest.entries.filter { $0.kind == .file }
        self.chunkSize = max(1, min(chunkSize, BulkFrame.maximumChunkSize))
    }

    deinit { try? handle?.close() }

    public func nextChunk() throws -> Data? {
        while index < files.count {
            if handle == nil {
                handle = try FileHandle(forReadingFrom: batch.fileURLs[index])
                bytesRead = 0
                hasher = SHA256()
            }
            if let data = try handle?.read(upToCount: chunkSize), !data.isEmpty {
                bytesRead += UInt64(data.count)
                guard bytesRead <= files[index].byteCount else {
                    throw FileTransferBatchError.sourceChanged(files[index].path)
                }
                hasher.update(data: data)
                return data
            }
            try handle?.close()
            handle = nil
            let digest = hasher.finalize().map { String(format: "%02x", $0) }.joined()
            guard bytesRead == files[index].byteCount, digest == files[index].sha256 else {
                throw FileTransferBatchError.sourceChanged(files[index].path)
            }
            index += 1
        }
        return nil
    }
}

public enum FileTransferBatchError: Error, Sendable, Equatable {
    case unsupportedItem(String)
    case sourceChanged(String)
}

public enum FileTransferStatus: String, Sendable, Codable {
    case queued
    case awaitingApproval
    case transferring
    case verifying
    case completed
    case rejected
    case canceled
    case failed
}

public struct FileTransferNameMapping: Sendable, Equatable, Codable {
    public let original: String
    public let local: String

    public init(original: String, local: String) {
        self.original = original
        self.local = local
    }
}

public struct FileTransferHistoryEntry: Sendable, Equatable, Codable, Identifiable {
    public var id: TransferID { transferID }
    public let transferID: TransferID
    public let name: String
    public let direction: TransferDirection
    public let totalBytes: UInt64
    public let deviceName: String
    public let startedAt: Date
    public var status: FileTransferStatus
    public let nameMappings: [FileTransferNameMapping]

    public init(
        transferID: TransferID,
        name: String,
        direction: TransferDirection,
        totalBytes: UInt64,
        deviceName: String,
        startedAt: Date,
        status: FileTransferStatus,
        nameMappings: [FileTransferNameMapping] = []
    ) {
        self.transferID = transferID
        self.name = name
        self.direction = direction
        self.totalBytes = totalBytes
        self.deviceName = deviceName
        self.startedAt = startedAt
        self.status = status
        self.nameMappings = nameMappings
    }
}

public struct FileTransferHistoryStore: Sendable, Equatable, Codable {
    public static let maximumCount = 100
    public private(set) var entries: [FileTransferHistoryEntry] = []

    public init() {}

    public mutating func record(_ entry: FileTransferHistoryEntry) {
        entries.removeAll { $0.transferID == entry.transferID }
        entries.insert(entry, at: 0)
        if entries.count > Self.maximumCount { entries.removeLast() }
    }

    public mutating func update(_ transferID: TransferID, status: FileTransferStatus) {
        guard let index = entries.firstIndex(where: { $0.transferID == transferID }) else { return }
        entries[index].status = status
    }

    public mutating func clear() { entries.removeAll() }
}
